src/ast/passes/semantic_analyser.cpp

```diff
@@ -2898,6 +2946,54 @@ void SemanticAnalyser::visit(AssignConfigVarStatement &assignment)
   Visit(assignment.expr);
 }
 
+void SemanticAnalyser::visit(VarDeclStatement &decl)
+{
+  const std::string &var_ident = decl.var->ident;
+
+  if (!IsValidVarDeclType(decl.var->type)) {
+    LOG(ERROR, decl.loc, err_)
+        << "Invalid variable declaration type: " << decl.var->type;
+  }
+
+  // This should be the first time we're seeing this variable
+  if (auto decl_search = variable_decls_[scope_].find(var_ident);
+      decl_search != variable_decls_[scope_].end()) {
+    if (decl_search->second != decl.loc) {
+      // Note this will change with proper block/lexical scoping
+      LOG(ERROR, decl.loc, err_)
+          << "Variable " << var_ident << " was already declared.";
+      LOG(ERROR, decl_search->second, err_) << "Initial declaration";
+    }
+  }
+
+  if (auto search = variables_[scope_].find(var_ident);
+      search != variables_[scope_].end()) {
+    if (!variable_decls_[scope_].contains(var_ident)) {
+      LOG(ERROR, decl.loc, err_)
+          << "Variable declarations need to occur before variable usage or "
+             "assignment. Variable: "
+          << var_ident;
+    } else if (is_final_pass()) {
+      // Update the declaration type if it was either not set e.g. `let $a;`
+      // or the type is ambiguous or resizable e.g. `let $a: string;`
+      decl.var->type = search->second.type;
+    }
+
+    if (is_final_pass() && !search->second.was_assigned) {
+      LOG(WARNING, decl.loc, out_)
+          << "Variable " << var_ident << " never assigned to.";
+    }
+
+    return;
+  }
+
+  bool can_resize = !(decl.set_type && decl.var->type.GetSize() > 0);
```

Comment:
Nit: De Morgan this to improve readability 
```suggestion
  bool can_resize = !decl.set_type || decl.var->type.GetSize() == 0;
```